rascaline/src/calculators/atomic_composition.rs

```diff
@@ -0,0 +1,269 @@
+    fn name(&self) -> String {
+        return "atom-centered composition features".into();
+    }
+
+    fn parameters(&self) -> String {
+        return serde_json::to_string(self).expect("failed to serialize to JSON");
+    }
+
+    fn keys(&self, systems: &mut [Box<dyn System>]) -> Result<Labels, Error> {
+        return CenterSpeciesKeys.keys(systems);
+    }
+
+    fn samples_names(&self) -> Vec<&str> {
+        if self.per_structure {
+            return vec!["structure"];
+        } else {
+            return vec!["structure", "center"];
+        }
+    }
+
+    fn samples(
+        &self,
+        keys: &Labels,
+        systems: &mut [Box<dyn System>],
+    ) -> Result<Vec<Arc<Labels>>, Error> {
+        assert_eq!(keys.names(), ["species_center"]);
+        let mut samples = Vec::new();
+        for [species_center_key] in keys.iter_fixed_size() {
+            let mut builder = LabelsBuilder::new(self.samples_names());
+
+            for (system_i, system) in systems.iter_mut().enumerate() {
+                if self.per_structure {
+                    builder.add(&[system_i]);
+                } else {
+                    let species = system.species()?;
+
+                    for (center_i, &species_center_sys) in species.iter().enumerate() {
+                        if species_center_key.i32() == species_center_sys {
+                            builder.add(&[system_i, center_i]);
+                        }
+                    }
+                }
+            }
+            samples.push(Arc::new(builder.finish()));
+        }
+
+        return Ok(samples);
+    }
+
+    fn supports_gradient(&self, parameter: &str) -> bool {
+        match parameter {
+            "positions" => true,
+            "cell" => true,
+            _ => false,
+        }
+    }
+
+    fn positions_gradient_samples(
+        &self,
+        keys: &Labels,
+        samples: &[Arc<Labels>],
+        systems: &mut [Box<dyn System>],
+    ) -> Result<Vec<Arc<Labels>>, Error> {
+        debug_assert_eq!(keys.count(), samples.len());
+        let mut gradient_samples = Vec::new();
+        for ([species_center_key], samples) in keys.iter_fixed_size().zip(samples) {
+            let mut builder = LabelsBuilder::new(vec!["sample", "structure", "atom"]);
+            if self.per_structure {
+                for (system_i, system) in systems.iter_mut().enumerate() {
+                    let species = system.species()?;
+
+                    for (center_i, &species_center_sys) in species.iter().enumerate() {
+                        if species_center_key.i32() == species_center_sys {
+                            builder.add(&[system_i, system_i, center_i]);
+                        }
+                    }
+                }
+            } else {
+                for (sample_i, [structure_i, center_i]) in samples.iter_fixed_size().enumerate() {
+                    builder.add(&[sample_i, structure_i.usize(), center_i.usize()]);
+                }
+            }
```

Comment:
I am bit puzzled here:

The code seems to work properly. But, the two logic branches give the same number of gradients with exactly the same sample names. Is this correct for the case of an atomic composition, that the gradients are the same regardless if I request a `per_atom` or `per_structure` descriptor?

If yes, we could remove the `self.per_structure==false` branch and simplify the code even further.

Reply:
Hmmm are you sure? Without experimenting with your code, I would expect the gradient samples to be structure, center, atom in the normal case and structure, atom in the per_structure case. Does it make sense to you? Can anybody confirm?

Reply:
Anyway, both gradients samples should be empty here. The positions/cell gradients of the composition are zero everywhere ...

Reply:
Se we do not initialize them at all?

Reply:
Yes we can return an empty set of samples

Reply:
Hmm but if I do something like

```rust
    fn positions_gradient_samples(
        &self,
        keys: &Labels,
        samples: &[Arc<Labels>],
        systems: &mut [Box<dyn System>],
    ) -> Result<Vec<Arc<Labels>>, Error> {

        debug_assert_eq!(keys.count(), samples.len());
        let gradient_samples = Vec::new();
        return Ok(gradient_samples);
    }
```

The tests for finite difference of course don't work anymore. So, I am not sure what you mean.

Reply:
Sorry, needs to be a vector of empty labels (one for each key). Something like this should work

```rs
let gradient_samples = Arc::new(Labels::empty(["sample", "structure", "atom"]));
return Ok(vec![gradient_samples; keys.count()]);
```

Reply:
Ahh I see, yes this should do the job.

Reply:
However, this causes problems when running `sum_over_samples`. This leads to an `ValueError: cannot reshape array of size 0 into shape (0,newaxis)` because the gradients are empty. Also I am not sure what happening when applying `join`.

Reply:
I would say any error like this is a bugs in `sum_over_samples`/`join`/... which I would rather fix there